### ISSUE TITLE
Comments on private imports.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#1767]
Collected comments on https://github.com/dart-lang/language/blob/master/working/modules/private-imports.md

All in all, I like the idea. I think it can work. (And I can't wait to get rid of `@VisibleForTesting`).

If the spec still uses "accessible" about whether you can see a private member, simply saying that a private declaration is accessible if it's privately imported might go a long way (but we probably need something more fine-grained to handle conflict resolution).

## individual imports/shows/hides

As for importing individual private members, not just all of them, that's probably wise, but I'd still want a "private import" marker on the import, and then allow `show`/`hide` with private names. I'd probably not distinguish between individual class members, you can `hide` or `show` the entire class/mixin/extension/whatnot.  (And you can have a separate non-private import of the same library to get the public API anyway).

Using `show _` as the marker is not great ... because you could have a declaration named `_`, and with individual `show`/`hide` modifiers, `show _` could be showing that. maybe `import private "lala.dart" show ...` is an option.

## Top-level declarations

For imported private top-level names, the specification could be that private declarations from the other library become *aliased* in the local library with library-local names of the importing library. That's ... probably not going to work for members, so we might as well bite the bullet and admit that we have foreign private names in scope. 
In that case, an unqualified library local variable which fails to resolve to something declared in the current library will then test against "private import" names before bailing out.

It does mean that `_foo` will check for imported private members *before* defaulting to `this._foo`, and adding a private member to a privately imported library can now be a breaking change. It can only be breaking inside the same package, so you should be able to detect it immediately and fix the problem. (Worst case: It doesn't break, just does something wrong).

## Member names 

A private member access is more complicated.

A static member or constructor access like `Foo._bar` is trivial, it can't be anything else than a private name declared in the library declaring `Foo`. So, that should be safe.

An instance member access, `o._foo`, is more complicated.
If `o` is typed as `dynamic`, I agree that it should always be a local library name lookup.

Otherwise, if `o` has multiple "accessible" private members named `_foo`, we have a conflict.

I'd say that if there is a `_foo` member in the interface of `o` declared in the *current* library, I think it should shadow foreign private members. If the foreign member is declared in a supertype, you can always upcast to that supertype (which doesn't have the local-library member) and then do the private member access.
If not, then it's just a compile-time error if there is a private `_foo` from more than one library in the type.

So: I think we should give local-library declarations preference. 

An extension member access, `o._foo(x)` needs to first check `o` for an "accessible" instance member, then check for available extension members if it didn't fine any instance members. Nothing revolutionary.
So, introducing a private member on a privately imported class can break something which currently invoke a local private extension method.
Then we just need to decide whether extension methods are "accessible" or not as normal.

Also doing an import of an *unnamed* private extension, like `extension <T> on T { ... }`, could make it accessible and not provide a way to hide it. Again, same package, so just give it a name and move on.

## Overriding private members.

I'd recommend *not* allowing that at all.
If you *declare* something, it will be private to your own library. You don't get to override other libraries' private members (that would also be potentially costly in optimization opportunities, we've done things to ensure that you can directly call private members that are not locally overridden).

So, private *read access*, but no ability to intrude into the library.

## Eliminating parts

Eliminating parts is ... laudable. Requires private imports in both directions (or n*n imports for n parts that all access each other, but that's likely rare.)

**Comment by leafpetersen** [comment]
> we've done things to ensure that you can directly call private members that are not locally overridden

I don't think this is true is it?  Because noSuchMethod forwarders can always implicitly override anything?

**Comment by munificent** [comment]
> I'd still want a "private import" marker on the import, and then allow `show`/`hide` with private names.

That's what I was thinking too.

> Using `show _` as the marker is not great ... because you could have a declaration named `_`, and with individual `show`/`hide` modifiers, `show _` could be showing that. maybe `import private "lala.dart" show ...` is an option.

I admit that `show _` is perhaps more cute than wise. But `import private` feels strange to me too because, as weird as this is, Dart doesn't use the word "private" anywhere else in its syntax. The only text in Dart itself that means "private" is `_`, which is how I got to that weird syntax. I'm not strongly attached to it, though.

> It does mean that `_foo` will check for imported private members _before_ defaulting to `this._foo`, and adding a private member to a privately imported library can now be a breaking change.

Good point. I think that's probably still the right behavior, though because it's always still going to be your code. (And, in practice, top-level lower case names that collide with instance members are very rare.)

> I'd say that if there is a `_foo` member in the interface of `o` declared in the _current_ library, I think it should shadow foreign private members. If the foreign member is declared in a supertype, you can always upcast to that supertype (which doesn't have the local-library member) and then do the private member access.

Ah, I like this. Updated the proposal.

> I'd recommend _not_ allowing [overriding] at all. If you _declare_ something, it will be private to your own library.

I'm not morally opposed to this restriction. But, in practice, I do think it will end up being useful for tests to be able to override private members in order to mock or stub stuff out, which is why I included it. I would be OK with removing this from the proposal if we think it's problematic.

Thank you for the feedback!

**Comment by lrhn** [comment]
>> we've done things to ensure that you can directly call private members that are not locally overridden
>
> I don't think this is true is it? Because noSuchMethod forwarders can always implicitly override anything?

I guess the things we can (and maybe do) do now are voided by simply being able to see the private class declarations, whether you override its members or not.

Currently a private class which isn't extended or implemented inside the declaring library, is guaranteed to not have its members overridden anywhere (because no-one can extend or implement it).
Calling any method on a receiver of that type inside the library can make the call on-virtual because of that.

The private import feature breaks that whether it allows overriding the members or not, because it allows *implementing* the private class outside of the declaring library, and (as you say) implement it using `noSuchMethod`.

(I still think it's safer and cleaner to not allow overriding private members from other libraries. If you can do that, we need to find a way to allow declaring multiple of members with the same private base-name, like `A::_foo() => 42;` and `B::_foo() => 42`. If you do need that, it should be possible.)

**Comment by munificent** [comment]
> The private import feature breaks that whether it allows overriding the members or not, because it allows _implementing_ the private class outside of the declaring library, and (as you say) implement it using `noSuchMethod`.

This is true. However, if we're willing to accept that a Dart compiler is aware of the contents of an entire package, then that's still sufficient to tell if a private member is overridden anywhere in the world and devirtualize it if not.

**Comment by natebosch** [comment]
> However, if we're willing to accept that a Dart compiler is aware of the contents of an entire package, then that's still sufficient to tell if a private member is overridden anywhere in the world and devirtualize it if not.

We can't guarantee that a compiler is aware of the full content of a package until we have the whole world available. When we do a modular compile in bazel there could other modules in the same package that aren't seen until later.

> because you could have a declaration named `_`,

I would be fine with disallowing the name `_` at the top level of a library. If these features are gated behind a language version this is also non-breaking.

**Comment by Silentdoer** [comment]
Since dart allows to define _ as a variable name(`int _ = 99`), I don’t think `show _` should be used

**Comment by lrhn** [comment]
If we remove the ability to declare variables named `_` at the same time, and, say, make it a non-binding pattern (so `foo(_) => 42` would declare a parameter with no name to refer to it by, and you could do `bar(_, _) => ...` to declare two unused parameters), then that would not be a problem. You wouldn't be allowed to have a top-level declaration without a name.

**Comment by Silentdoer** [comment]
> 
> 
> If we remove the ability to declare variables named `_` at the same time, and, say, make it a non-binding pattern (so `foo(_) => 42` would declare a parameter with no name to refer to it by, and you could do `bar(_, _) => ...` to declare two unused parameters), then that would not be a problem. You wouldn't be allowed to have a top-level declaration without a name.

So dart will not support static imports, will it? such as:
```dart
// a.dart
class Foo {
  void _(){print('ss);}
}
// b.dart
static import './b.dart' show Foo._;

void main() {
_();  // prints ss
}
```

**Comment by munificent** [comment]
> So dart will not support static imports, will it?

We don't have any plans for that currently, no, though we have discussed it.
